### ISSUE TITLE
Interfacing extended XYZ readers and writers for AtomsBase

### ISSUE THREAD

**Issue by jameskermode** [JuliaMolSim/AtomsBase.jl#40]
Hello! I'd like to contribute [extended XYZ](https://wiki.fysik.dtu.dk/ase/ase/io/formatoptions.html#extxyz) readers and writers to AtomsBase. I have an existing Julia package, [ExtXYZ.jl](https://github.com/libAtoms/ExtXYZ.jl), which calls a [C library](https://juliahub.com/ui/Packages/extxyz_jll/89uRR) containing parser that is auto-generated from a [formal grammar](https://github.com/libAtoms/extxyz) for the extended XYZ format that we've been working on recently, as well as a C writer.

There are two things I'd like to discuss before starting work ob this:

- **Package Dependencies**. ExtXYZ.jl is already interfaced to @cortner's JuLIP.jl. There we made JuLIP depend on ExtXYZ, but since it seems AtomsBase is designed to contain the basic interface only I'm wondering if it would be better for me to make my ExtXYZ package depend on AtomsBase instead? Or perhaps there should be a new package which uses both of them?

- **Arbitrary Data**. Extended XYZ format allows for arbitrary per-atom and per-config data. This is useful e.g. to store (perhaps multiple sets of) energies and forces associated with a structure when training machine learning potentials, or for visualisation of per-atom quantities with [OVITO](https://www.ovito.org) which can also read extxyz format. In ASE these are stored in the `Atoms.arrays` and `Atoms.info` dictionaries, respectively. In JuLIP we did something similar, with a single `data` dictionary. I see there is a `data` dict in the `Atom` type, but from what I can tell storing additional data associated with each structure is not included in the abstract `AbstractSystem` API. Is this something you've discussed already and ruled out? If so should I add interface routines which convert to/from `(AtomsBase.AbstractSystem, Dict)` pairs and the dictionaries that ExtXYZ works with? Or would it be better to define my own concrete subtype of AbstractSystem which includes e.g. `per_atom_properties::Dict{Symbol,Any}` and `per_structure_properties::Dict{Symbol,Any}` dictionaries, or similar?

**Comment by rkurchin** [comment]
Thanks for your interest in supporting AtomsBase, this is awesome!

To respond to your first point: our general vision is that other packages would depend on AtomsBase. The goal is for this package to stay very lightweight and provide a minimal set of functions that allow easy "translation" between different structure representations.

To your second point, there are two general "paradigms" by which you can introduce this dependency: by either using one of our two built-in system types (for an example of this very analogous to what you're doing, check out [AtomIO.jl](https://github.com/mfherbst/AtomIO.jl)), or by defining a new system type and dispatching the necessary interface functions on it. Given your desire to support these extra data fields, I think this second approach (as you allude to) makes the most sense in the immediate term.

THAT BEING SAID, we have definitely discussed the possibility of something like this at the level of the interface, but hadn't (yet) converged upon a solution we felt was generic enough. So we would definitely be open to discussing introducing new interface functions to access these more "arbitrary" data fields in the future, but would want to get input from other folks to make sure that the way that we do it is sufficiently general.

@mfherbst and/or @jgreener64, feel free to chime in with further opinions! I suspect @cortner may have thoughts as well. :)

**Comment by cortner** [comment]
@jameskermode maybe we can play with a `FlexAtom` struct building on the `ACE.State` implementation. This would allow for fully flexible fields, we could even distinguish static and dynamic fields, but I'm a bit worried about dynamic fields re performance. This struct could then implement the `AtomsBase` interface. And for us it would be the first step towards retiring JuLIP and starting to incorporate the JuLIP functionality into other packages.

**Comment by rkurchin** [comment]
This probably veers towards "broader discussion" and might belong in another issue, but another related package that would definitely need support for "arbitrary data" type of fields (see e.g. [this list](https://cclib.github.io/data.html)) is [Cclib.jl](https://github.com/schneiderfelipe/Cclib.jl).

**Comment by mfherbst** [comment]
I'm starting to collect these ideas at https://github.com/mfherbst/AtomIO.jl/issues. Feel free to add more packages if you know them.

**Comment by jameskermode** [comment]
Thanks all for the encouraging responses. I’ll give @cortner’s suggestion of defining a flexible Atoms struct similar to `ACE.State` but supporting the `AtomsBase` interface a go within ExtXYZ.jl in the first instance and report back here.

I came up with the arbitrary per-atom and per-config data paradigm that the extended XYZ format represents in about 2005 and it’s served us remarkably well since then, so I’d be an advocate of finding a way to fit this into the `AbstractSystem` if it can be done without compromising efficiency.

In ExtXYZ, per-atom quantities can be scalar or vector, while per-config quantities can be scalar, vector or matrix. This allows many quantities including total charge, forces and energies, dipole moment, atomic stresses to be represented. It would potentially be nice to be even more general than this and allow e.g. per-atom matrix or tensor quantities as well.

**Comment by cortner** [comment]
@jameskermode - do you mean you want to just try locally whether my approach is sufficiently flexible for XYZ? Otherwise we could collaborate on a separate package that implements the `Atom` and `System` interfaces.

**Comment by cortner** [comment]
James - why not publish this as a separate package independent of XYZ? We could start making ACE compatible with that, and then try it out with Molly etc? If all this goes well then we can start thinking about which parts of JuLIP to get rid of.

**Comment by jameskermode** [comment]
Happy to do that. Any good name suggestions?

**Comment by cortner** [comment]
Can this be closed?

**Comment by jameskermode** [comment]
Yes, ok for me to close
